goodpractice tech note (#1307)
ropensci/roweb3: content/blog/2026-06-01-goodpractice/index.md

---
slug: goodpractice

> title: "Our goodpractice package has new superpowers"

Reviewer: ```suggestion
title: "Our goodpractice Package Has New Superpowers"
```

author:
    - Mark Padgham

>     - Athanasia Mo Mowinckel

Reviewer: ```suggestion
    - Athanasia Mo Mowinckel
editor:
    - Steffi LaZerte
```

I have found myself writing my own code, but asking claude to commit them (making it look like claude wrote the code, but I don't care) because the commit messages are just so much better than what I write.

### From Mark

The entire idea for this major update was initiated and driven by Mo.
It was also my first real foray into using generative AI tools directly in rOpenSci packages, and I learnt a lot.
After I rejected the initial monster PR, things settled down into much smaller and manageable PRs that were always focussed on one specific thing.

From that point, Mo was effectively the coder (assisted by Claude), and I was the reviewer.
I never knew how much of the actual code was typed versus machine-generated, but was generally pleasantly surprised that I felt no need to inquire.
Nearly every PR was obviously going to improve the package, and many of them were genuinely creative ideas.
(That's where I disagree slightly with Mo's general description of the process: I am convinced a lot of the work behind this update came directly from her brilliant insights and experience, rather than her passive description of looking over pre-existing PRs.)

> The process started in Feb 2026, and involved [70 pull requests](https://github.com/ropensci-review-tools/goodpractice/pulls?q=sort%3Aupdated-desc%20is%3Apr%20is%3Aclosed%20author%3Adrmowinckels).

Reviewer: !!!!

So many of these really were independent and new ideas.
That enabled me to view each one with fresh eyes, and to think about whether I might approach anything in different ways.
Over so many PRs, there was a lot of back-and-forth, from pushing back on Claude insisting on nested loops, to figuring out whether code was best parsed and analysed as text, or as a syntax tree.
As Mo indicated, a lot of those discussions ultimately converged towards us bringing the power of treesitter in to the package, and even helping to replace previous text-based code checks with more efficient and accurate AST approaches.

Our collaboration was very pleasant and enriching, and really felt like a direct cooperation between Mo as the ideas factory, me as the reviewer, and Claude as nothing more than a mediator of our ideas.
The key role of Claude throughout the process was in handling all of the fiddly, technical details.

> We both looked carefully at every single line of code, but the use of Claude enabled our conversations to remain at higher, conceptual levels than what would have happened if we had to actually do all the technical implementation ourselves.

Reviewer: ```suggestion
We both looked carefully at Claude's code, but the use of Claude enabled our conversations to remain at higher, conceptual levels than what would have happened if we had to actually do all the technical implementation ourselves.
```

Author: That one's a rare rejection from me @steffilazerte, because I don't like the close repetition of "Claude", and so prefer the way it is.

#>
#>     R/utils.R:3
#>     R/utils.R:73
#>
#> ✖ define exported (user-facing) functions before internal helper functions within each R source file.
#>
#>     R/api.R:85
#>
#> ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

---

## Notes on the development process

### From Mo (Athanasia)

I'll be honest with you: a lot of this contribution was done with an AI assistant (Claude Code using Opus 4.5) at my elbow. 
I want to say something about that, because I think the honest version is more useful than the marketing version.
The way Mark approached reviewing this work was very thoughtful and helpful. 

Based on his feedback and how it was structured, I've learned a lot about what a good code review looks like, and it's been absolutely wonderful.
As a person who has mainly worked on their own code-bases alone, I have adopted a style of working that often creates diff monsters (hundreds of files changed), which can make reviewing very hard.
Indeed, my first PR to this project was such a monster, and Mark pushed back asking for several smaller PRs he could actually manage to review. 

> This is also where I learned about [git worktrees](https://git-scm.com/docs/git-worktree) (Thank you [Maëlle](/author/ma%C3%ABlle-salmon/)!) and asking Claude to split the work up into several worktrees for easier review, was actually the first substantial change to how I now approach working with an AI assistant. 

Reviewer 2: ```suggestion
This is also where I learned about [git worktrees](https://git-scm.com/docs/git-worktree) (Thank you [Maëlle](https://masalmon.eu/2026/04/23/git-worktree-again/)!) and asking Claude to split the work up into several worktrees for easier review, was actually the first substantial change to how I now approach working with an AI assistant. 
```

Checks were identifiable with the single function, [`all_checks()`](https://docs.ropensci.org/goodpractice/reference/all_checks.html).
The only way to control which checks were run was to pass a `checks` parameter to the main function as a character vector of the names of checks you wanted to run.
So, for example, if you wanted to skip the [rcmdcheck](https://rcmdcheck.r-lib.org) checks, you had to do something like this:
``` r
mychecks <- all_checks()[which(!grepl("^rcmdcheck", all_checks()))]
gp(checks=mychecks)
```
That wasn't easy!

This update finally makes it easier to control checks, by defining them in "groups", which generally corresponding to the separate packages which run the actual checks.
The groups can be seen with the new [`all_check_groups()` function](https://docs.ropensci.org/goodpractice/reference/all_check_groups.html), and controlled with the [`checks_by_group()` function](https://docs.ropensci.org/goodpractice/reference/checks_by_group.html).
Checks can be controlled by simply naming the check groups you want to run, like this:

> ``` r

Reviewer 3: ```suggestion

``` r
```

Reviewer 3: add empty line before code chunk

The number of checks has been increased from 230 to 338, with all listed in the output of the [`all_checks()` function](https://docs.ropensci.org/goodpractice/reference/all_checks.html).

## Improved check reporting

The goodpractice package has always been designed for console output.
This updated version now provides immediate and consistent detail on check progress while running.
The [cli package](https://cli.r-lib.org) is also used to give consistently formatted output for all check groups.

The large increase in numbers of checks also means that directly printing goodpractice results may fill several screens of output.
This update also adds a `groups` parameter to the [`print()` method](https://docs.ropensci.org/goodpractice/reference/print.goodPractice.html), to enable printing check results only for specified groups.
So you can run the main [`gp()` function](https://docs.ropensci.org/goodpractice/reference/gp.html) once, then step through each check group by printing results only for that group, fixing those, and then moving on to the next.
We've found this turns what could feel like an overwhelming wall of output into a manageable to-do list.

> ## Let us know what you think

Reviewer 2: I think it'd be neat to have one example or more in this post.

Author: Good idea; done now.

> 

Reviewer 2: I also thought the post would go a bit more into:

- @drmowinckels' now being an author of the package, and the push behind many of these improvements;
- the use of AI for development.

A few sentences for each would go a long way.

Author: @drmowinckels What are your thoughts on this? My notion was to publish this as a first tech note, and then you could lead a second blog post describing how you actually went about it? If you can't be bothered writing all that up, we can easily just add a few notes here. Up to you :+1:

Reviewer 3: I think maybe adding it here would make more sense. I can add some suggestions asap regarding that from my perspective and you can adapt using your perspective?

Author: That'd be great - thank you!

Reviewer: Reading Mo's viewpoint suggestion above, my suggestion is to add a brief sentence to the paragraph on maintainership to note that Mark is the maintainer, and Mo is now an author after substantial contributions to this new version. Then I think Mo's thoughts on the AI workflow and review would actually be better as a second small post talking about the process. That being said, if you'd both rather keep it here, that's not wrong :D but I would still add the explicit note about contributions to the start.

## Improved check reporting

The goodpractice package has always been designed for console output.
This updated version now provides immediate and consistent detail on check progress while running.
The [cli package](https://cli.r-lib.org) is also used to give consistently formatted output for all check groups.

The large increase in numbers of checks also means that directly printing goodpractice results may fill several screens of output.
This update also adds a `groups` parameter to the [`print()` method](https://docs.ropensci.org/goodpractice/reference/print.goodPractice.html), to enable printing check results only for specified groups.
So you can run the main [`gp()` function](https://docs.ropensci.org/goodpractice/reference/gp.html) once, then step through each check group by printing results only for that group, fixing those, and then moving on to the next.
We've found this turns what could feel like an overwhelming wall of output into a manageable to-do list.
Like this:
``` r
x <- gp()
print(x, "description")
#> [1] Aww! Shining package! Keep up the priceless work!
```

> 

Reviewer 3: ```suggestion

## A note from Mo (Athanasia) on the process

I'll be honest with you: a lot of this contribution was done with an AI assistant (Claude Code using Opux 4.5) at my elbow. 
I want to say something about that, because I think the honest version is more useful than the marketing version.
The way Mark approached reviewing this work was very thoughtful and helpful. 

Based on his feedback and how it was structured, I've learned alot about how a good code review can look like, and its been absolutely wonderful.
As a person who has mainly worked on their own code-bases alone, I have adopted a style of working that often creates diff monsters (hundreds of files changed), which would make reviewing very hard.
Indeed, my first PR to this project was such a monster, and Mark pushed back asking for several smaller PRs he could actually manage to review. 
This is also where I learned about [git worktrees](https://git-scm.com/docs/git-worktree) (Thank you [Maëlle](https://ropensci.org/author/ma%C3%ABlle-salmon/)!) and asking Claude to split the work up into several worktrees for easier review, was actually the first substantial change to how I now approach working with an AI assistant. 
Through this collaboration I learned a lot about splitting my work into better sizes and chunks.

The shape of it was something like this:

I'd locate an issue I thought was meaningful to tackle, and provide Claude Code with the issue context and possibly an idea of how to solve it.
Once it had a solution, I'd review it, push back where it felt off, and we'd iterate. 
Mark would then review the PR and point out the things I'd missed. 
Three-way collaboration, more or less, with the AI doing the typing and me doing the deciding.      

What worked well was scaffolding. 
Drafting a new check, wiring up the prep step that goes with it, generating the test cases for the edge paths I'd otherwise forget — that kind of work compresses really nicely with an AI. 
I would often ask it to generate some tests first, then create code that would pass the tests.
This way we had a clear idea of what we wanted the new code to do, and then solve it. 

What worked less well was anything involving judgement I hadn't articulated yet. 
The first few times Claude opened a  PR for me, it left the test-plan checkboxes unticked — even when the tests already passed. 
That's not wrong exactly, but it's misleading to a reviewer. 
I had to say "no, check the box if the test exists and passes; an unchecked box reads as TODO." 
Similarly, when Mark left inline code suggestions on a PR, Claude tried to helpfully re-implement them locally — which would have stripped Mark's attribution off his own contribution. 
We had to agree: suggestions get accepted on GitHub, not retyped.

And then there's the unglamorous part. 
It would often forget instructions, despite having them documented in agents.md and local memory for the project. 
These were mostly trivial, but at times fairly bad.
Some of the first PRs we made, Mark pointed out that the solutions were using regexp rather than AST.
I have to admit, I didn't really know about AST before starting this project, and I am so glad I know of it.
Now, despite being told very clearly that we wanted AST solutions, it would often forget and implement regexp based solutions. 
It would also very often revert to using for-loops rather than vectorization, and creating nested for-loops into the 3rd or 4th level — which is just horrible to follow as a human.

If you take one thing from this aside, take this: AI-assisted contribution didn't mean handing the package over to a machine. 
It meant I could move faster on the parts I already understood, explore more confidently on the parts I didn't, and spend my actual attention on the decisions that mattered — what to check for, how to group it, what to call it, and what to leave out. 
The friction taught me where to watch closely. 
The speed-ups taught me where I could trust the loop.

It's a different way of working, and I'm still figuring out the shape of it, but it has enabled me to contribute to this project despite some very severe health issues making it hard for me to work as normal.

And one thing Claude always does better than me: writing good commit messages.
I have found myself writing my own code, but asking claude to commit them (making it look like claude wrote the code, but I don't care) because the commit messages are just so much better than what I write.

```

Author: Thanks @drmowinckels - I thought about setting up a new post, but actually think this could work well here. The current post is fairly short, and it might make it both more interesting and more coherent to have all together. I'll commit your suggestion tomorrow, and edit from there.

Be warned that I'm going to reword quite dramatically, because i think you're being way too modest there :muscle: It really reads as if you just pointed claude at open issues, and guided it from there. But I was on the receiving end, and the vast, vast majority of your inputs really were *ab initio* brilliant ideas from you, borne of your wisdom and insights into coding practices in general. Hopefully you'll be available tomorrow for a bit of back-and-forth. Thanks for everything!
